### PR TITLE
feat: Assign colors to events

### LINKED ISSUE #7507: [CAL-2791] Assign colors to events to make them easier to find in the booking overview
Would be cool to be able to assign different colors to individual events so that they can be better visually assigned in the overview of booked events (see Calendly).

image

image

By the way: Categories for the event types would also be great. This way you can bring more order into it, if you offer many different types of events, like me.

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-2791](https://linear.app/calcom/issue/CAL-2791/assign-colors-to-events-to-make-them-easier-to-find-in-the-booking)</sub>

### PR DESCRIPTION
**Pull request by anikdhabal**
## What does this PR do?

- Fixes #7507
- Fixes CAL-2791

Screenshot 2024-06-03 232921
Screenshot 2024-06-03 232859

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected)
- [ ] I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com)
- [ ] I have added or modified automated tests that prove my fix is effective or that my feature works (PRs might be rejected if logical changes are not properly tested)

## How should this be tested?

- Are there environment variables that should be set?
- What are the minimal test data to have?
- What is expected (happy path) to have (input and output)?
- Any other important info that could help to test that PR

## Checklist

- I haven't read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- My code doesn't follow the style guidelines of this project
- I haven't commented my code, particularly in hard-to-understand areas
- I haven't checked if my changes generate no new warnings

### CODE REVIEW

File: apps/web/components/eventtype/EventAdvancedTab.tsx [hunk 1]
```diff
@@ -532,6 +546,49 @@ export const EventAdvancedTab = ({ eventType, team }: Pick<EventTypeSetupProps,
...
+                    setColourError(false);
+                    formMethods.setValue("eventTypeColour", value, { shouldDirty: true });
+                  } catch (err) {
+                    setColourError(true);
+                  }
+                }}
+              />
+              {colourError ? (
+                <div className="mt-4">
+                  <Alert severity="warning" message={t("light_theme_contrast_error")} />
```
Comment (Amit91848) [inline apps/web/components/eventtype/EventAdvancedTab.tsx:585]: > Light Theme color doesn't pass contrast check. We recommend you change this colour so your buttons will be more visible.

Ig we should change this text. We aren't talking about button's visibility here.

**Review by Amit91848 — DISMISSED** [review]
Amazing work so far @anikdhabal ! Found a small issue that should be fixed. Can you also fix the merge conflicts.

You select a specific colour for event type and save the event type. Now you toggle the colour settings off and on again, but don't save the event type. The colour reverts to its default value, becoming inconsistent with current state of event type. Ideally it should still be the colour previously saved as we did not make any changes to it. Attached a video for better understanding.

https://github.com/calcom/cal.com/assets/74371312/478b97c9-0bf5-42fc-be15-3f940a3c213d

File: apps/web/public/static/locales/en/common.json [hunk 2]
```diff
@@ -2486,6 +2486,8 @@
   "unable_to_subscribe_to_the_platform": "An error occurred while trying to subscribe to the platform plan, please try again later",
   "updating_oauth_client_error": "An error occurred while updating the OAuth client, please try again later",
   "creating_oauth_client_error": "An error occurred while creating the OAuth client, please try again later",
+  "event_type_colour": "Event type colour",
```
Comment (CarinaWolli) [inline apps/web/public/static/locales/en/common.json:2489]:
```suggestion
  "event_type_color": "Event type color",
```
We should use American English

**Review by CarinaWolli — COMMENTED** [review]
We want to use american english so `color` and not `colour`. But feature looks great 🙌🏻

**Review by Udit-takkar — DISMISSED** [review]
This doesn't match the designs in figma https://www.figma.com/design/UajFu4M1APhn2AywAEJkJr/New-Features?node-id=4673-355405&t=ih0GY2rAHb4Wc3wX-0.

This design is marked as "ready for dev"

Screenshot 2024-08-01 at 12 40 27 AM

**Comment by anikdhabal** [comment]
> This doesn't match the designs in figma https://www.figma.com/design/UajFu4M1APhn2AywAEJkJr/New-Features?node-id=4673-355405&t=ih0GY2rAHb4Wc3wX-0.
> 
> This design is marked as "ready for dev"
> 
> Screenshot 2024-08-01 at 12 40 27 AM

Hey @Udit-takkar , why do we need to create a different color picker component with limited colors when we already have one? I think the current picker provides more color options. Additionally, creating a new component would make the code larger, and the use case for that color component is only for the event type color picker. Lmk what you think

File: apps/web/public/static/locales/en/common.json (hunk 2 above)
Comment (anikdhabal) [reply apps/web/public/static/locales/en/common.json]: done

**Comment by Udit-takkar** [comment]
> Hey @Udit-takkar , why do we need to create a different color picker component with limited colors when we already have one? 

We can't allow any colour to be selected as we have to make sure it doesn't fail the WCAG Contrast guidelines https://developer.mozilla.org/en-US/docs/Web/Accessibility/Understanding_WCAG/Perceivable/Color_contrast. Just confirm with @ciaranha once about the design and let me know

**Comment by anikdhabal** [comment]
> > Hey @Udit-takkar , why do we need to create a different color picker component with limited colors when we already have one?
> 
> We can't allow any colour to be selected as we have to make sure it doesn't fail the WCAG Contrast guidelines https://developer.mozilla.org/en-US/docs/Web/Accessibility/Understanding_WCAG/Perceivable/Color_contrast. Just confirm with @ciaranha once about the design and let me know

Sure @Udit-takkar , and for WCAG contrast guidelines, we can add a check and display a message. I think in our code, we check the contrast for both dark and light modes in some components

**Comment by anikdhabal** [comment]
> We can't allow any colour to be selected as we have to make sure it doesn't fail the WCAG Contrast guidelines https://developer.mozilla.org/en-US/docs/Web/Accessibility/Understanding_WCAG/Perceivable/Color_contrast. Just confirm with @ciaranha once about the design and let me know

Ciaran confirmed that we can use the color picker that we should have

**Review by sean-brydon — CHANGES_REQUESTED** [review]
CleanShot 2024-08-08 at 11 55 49
Translation not loading
--
CleanShot 2024-08-08 at 11 57 23
I was able to set this colour and didnt get an error

CleanShot 2024-08-08 at 11 59 50

Created a [replit](https://replit.com/@seanbrydon4/wcagchecker#index.ts) for easier testing of this fn manually.

We may honestly need a light/dark mode colour selector if we're not going for a preset color pallets like in the designs.

**Comment by anikdhabal** [comment]
@sean-brydon Thanks for the review. I'm on it. Also, the translation is perfect locally on my end. I don't know why it's not loading on your end.

**Comment by anikdhabal** [comment]
Hey @sean-brydon, the problem is that the way we handle the contrastCheck function is not optimal. The function is guaranteed to return a boolean, but we unnecessarily use it inside a try-catch block in every instance.

Before:-
Screenshot 2024-08-10 133818

After changes:-

Screenshot 2024-08-10 135243

**Review by sean-brydon — DISMISSED** [review]
Nice one! Thanks for making them changes - much better behaviour now

**Review by sean-brydon — APPROVED** [review]
Tested again - this is a neat feature!

**Pull request merged**